include/exadg/operators/operator_base.cpp

```diff
@@ -306,6 +360,68 @@ OperatorBase<dim, Number, n_components>::apply_add(VectorType & dst, VectorType
   }
 }
 
+template<int dim, typename Number, int n_components>
+void
+OperatorBase<dim, Number, n_components>::apply_matrix_based(VectorType &       dst,
+                                                            VectorType const & src) const
+{
+  if(this->data.sparse_matrix_type == SparseMatrixType::Trilinos)
+  {
+#ifdef DEAL_II_WITH_TRILINOS
+    if constexpr(std::is_same_v<Number, double>)
+    {
+      system_matrix_trilinos.vmult(dst, src);
+    }
+    else
+    {
+      // create temporal vectors of type double
+      dealii::LinearAlgebra::distributed::Vector<double> dst_double, src_double;
+      dst_double.reinit(dst, false);
+      src_double.reinit(src, true);
+      src_double = src;
+
+      system_matrix_trilinos.vmult(dst_double, src_double);
+
+      // convert: double -> Number
+      dst.copy_locally_owned_data_from(dst_double);
+    }
+#endif
+  }
+  else if(this->data.sparse_matrix_type == SparseMatrixType::PETSc)
+  {
+#ifdef DEAL_II_WITH_PETSC
+    if(system_matrix_petsc.m() > 0)
+    {
+      apply_petsc_operation(dst,
+                            src,
+                            petsc_vector_dst,
+                            petsc_vector_src,
+                            [&](dealii::PETScWrappers::VectorBase &       petsc_dst,
+                                dealii::PETScWrappers::VectorBase const & petsc_src) {
+                              system_matrix_petsc.vmult(petsc_dst, petsc_src);
+                            });
```

Comment:
@peterrum, @kronbichler This code compiled without float/double conversion as needed for the Trilinos case. Does this mean that the PETSc case computes the mat-vec product in float, or is the float/double conversion just done at a different place (e.g. within deal.II)?

Reply:
PETSc is not templated on the number type, so it is (in the standard compilation of PETSc) double-precision numbers in the matrix. I say standard compilation because you can configure PETSc with `--with-precision=single`, albeit it has been a long time ago I tried this the last time so I cannot say what precisely it entails today. In consequence, the number types get converted as we copy in/out of the PETSc vector, e.g. here: https://github.com/exadg/exadg/blob/c62fd0a241beab24a91cea6fd64680cbfb2c17cd/include/exadg/solvers_and_preconditioners/utilities/petsc_operation.h#L66-L69

Reply:
The `vmult()` function for AMG preconditioners is implemented in the class `PreconditionerAMG` used by both Trilinos/PETSc AMG preconditioners. So should we distinguish the implementation of this function for the Trilinos case and the PETSc case and do the float->double conversion of the deal.II vectors only in the Trilinos case, while we only call `amg->vmult()` for PETSc?

Reply:
Yes, in fact the conversion should not be necessary for PETSc, as we do an additional copy operation in the `petsc_operation` function.